src/core/lib/security/transport/security_connector.c

```diff
@@ -511,14 +537,107 @@ static void ssl_channel_add_handshakers(grpc_exec_ctx *exec_ctx,
           exec_ctx, tsi_create_adapter_handshaker(tsi_hs), &sc->base));
 }
 
+static const char **fill_alpn_protocol_strings(void) {
+  const size_t num_alpn_protocols = grpc_chttp2_num_alpn_versions();
+  const char **alpn_protocol_strings =
+      gpr_malloc(sizeof(const char *) * num_alpn_protocols);
+  size_t i;
+  for (i = 0; i < num_alpn_protocols; i++) {
+    alpn_protocol_strings[i] = grpc_chttp2_get_alpn_version_index(i);
+  }
+  return alpn_protocol_strings;
+}
+
+/* Attempts to replace the server_handshaker_factory with a new factory using
+ * the provided grpc_server_credentials. Should the credentials not be of type
+ * SSL or should the new factory creation fail, the existing factory will not be
+ * replaced. */
+static void replace_server_handshaker_factory(
+    grpc_ssl_server_security_connector *sc,
+    grpc_server_credentials *server_creds) {
+  const size_t num_alpn_protocols = grpc_chttp2_num_alpn_versions();
```

Comment:
nit: since `fill_alpn_protocol_strings()` function on its own determines how many protocol strings to fill in, the `num_alpn_protocols` should be something that that function returns, e.g.:
```
size_t num_alpn_protocols = 0;
const **alpn_protocol_strings = fill_alpn_protocol_strings(&num_alpn_protocols);
```

Reply:
Good idea. Will update.